cpp_guidelines/guidelines.md

```diff
@@ -0,0 +1,496 @@
+which namespace a closing `}` belongs.
+
+**Enforced by:**
+  * `.clang-format`: `FixNamespaceComments: true`
+
+**Example:**
+```c++
+namespace bad_example
+{
+}
+
+namespace good_example
+{
+} // good_example
+```
+
+### Put nested namespaces on separate lines.
+
+If available the C++17 syntax is preferred: `namespace first::second`.
+
+**Reason**: Each statement belongs on its own line. This includes namespace
+declarations.
+
+**Enforced by:**
+  * `.clang-format`: `CompactNamespaces: false`
+
+**Example:**
+```c++
+// bad example
+namespace foo { namespace bar {
+}}
+
+// good example
+namespace foo
+{
+    namespace bar
+    {
+    }
+}
+
+// best example
+namespace foo::bar
+-{
+}
+```
+
+### Put constructor initializers and inheritance list members on separate lines.
+
+**Reason:** It is easier to read multiple entries one-by-one than having them
+arbitrarily spread across lines.
+
+**Exception**: All initializers / list members fit into one line.
+
+**Enforced by:**
+  * `.clang-format`: `ConstructorInitializerAllOnOneLineOrOnePerLine: true`
+
+**Example:**
+```c++
+class foo
+{
+    foo()
+        : aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa{val}
+        , bbbbbbbbbbbbb{another}
+        , ccccccccccccccccccccccccccc{value}
+    {}
+};
+```
+
+## Indentation and alignment.
+
+### Use the Allman indentation style.
+
+**Reason:** This gives the code a nice vertical structure and optically
+separates different scopes from another. The K&R indentation style
+unnecessarily increases code density, making it harder to follow its
+structure.
+
+**Enforced by:**
+  * `.clang-format`: `BreakBeforeBraces: Allman`
+
+**Example:**
+```c++
+// bad example
+if(foo) {
+    /* something */
+}
+
+// good example
+if(bar)
+{
+    /* something */
+}
+```
+
+### Use spaces, never tabs.
+
+**Reason:** Spaces are the same on all machines, tabs are not.
+
+**Enforced by:**
+  * `.clang-format`: `UseTab: Never`
+
+### The indentation width is set to four columns.
+
+**Reason**: 80 (the maximum line width) is a multiple of four. Two columns do 
+not provide enough whitespace for visual structuring, eight columns (and
+above) fill the 80 columns too fast.
+
+**Enforced by:**
+  * `.clang-format`: `IndentWidth: 4`
+  * `.clang-format`: `TabWidth: 4`
+
+### Align after open brackets.
+
+**Reason:** This aligns multiple parameters nicely that do not fit into a
+single line. This rule applies to `()`, `<>` and `[]`.
+
+**Enforced by:**
+  * `.clang-format`: `AlignAfterOpenBracket: Align`
+  * `.clang-format`: `AllowAllArgumentsOnNextLine: false`
+  * `.clang-format`: `AllowAllConstructorInitializersOnNextLine: false`
+  * `.clang-format`: `AllowAllParametersOfDeclarationOnNextLine: false`
+
+**Example:**
+```c++
+auto some_long_function(argument1,
+                        argument2);
+```
+
+### Align consecutive assignments, bitfields and macros.
+
+**Reason:** It is visually pleasing and there is no reason against it.
```

Comment:
I started out aligning everything as well, but I ran into cases where it was not visually pleasing:
```
for () {
  ...
  someLongVariableName += 1;
  result                = thisIsAFunctionCall(withSomeArguments, andAnother);
  i                    += offset;
}
```
It happens ;) So I started to be more defensive here. But find out yourself ;)

Reply:
I am not very emotional on this one, as it does not happen often, but to me such alignments feel very bad.

Reply:
Alignments are bad because they are not invariant under refactoring, i.e. if you change the length of the longest variable name (or add a new longer one) all the others need to change, too.

In addition tom this, I think, prescribing alignment is only an option if we could configure a pre-commit hook running a fixer. I would not want to spent half my time coding aligning things, or fighting with my editor if it is configured to do this automatically.

Reply:
> Alignments are bad because they are not invariant under refactoring, i.e. if you change the length of the longest variable name (or add a new longer one) all the others need to change, too.

I totally agree. Alignment leads to unnecessarily large diffs. Especially with short line length, parameters to functions, templates, etc. are ofteh on multiple lines. If you rename a function, template, etc., then all the parameters in all invocations have to be realigned as well.

Reply:
I just tried the suggested .clang-format file on LLAMA. I am not sure if I like this one:
```
    constexpr std::size_t hardwareThreads = 2; // relevant for OpenMP2Threads
    using Distribution                    = common::ThreadsElemsDistribution<Acc, BLOCK_SIZE, hardwareThreads>;
    constexpr std::size_t elemCount       = Distribution::elemCount;
    constexpr std::size_t threadCount     = Distribution::threadCount;
    constexpr FP ts                       = 0.0001;
```
This one turned out nice:
```
constexpr auto DEFAULT_IMG_X   = 4096; /// width of the default image if no png is loaded
constexpr auto DEFAULT_IMG_Y   = 4096; /// height of the default image if no png is loaded
constexpr auto KERNEL_SIZE     = 8;    /// radius of the blur kernel, the diameter is this times two plus one
constexpr auto CHUNK_SIZE      = 512;  /// size of each chunk to be processed per alpaka kernel
constexpr auto ELEMS_PER_BLOCK = 16;   /// number of elements per direction(!) every block should process
```
I do not like this one:
```
        int x               = 0;
        int y               = 0;
        int n               = 3;
        unsigned char* data = stbi_load(argv[1], &x, &y, &n, 0);
```

I feel like I am weakly against alignment. Also the reasons given by others, like larger diffs, make me feel aligning is not a good idea.

Reply:
Count myself in against aligning for the above given reasons.

current (align) | No aligning of consecutive assignments
-- | --
0 | 5